syntax = "proto3";

package swift.zone;

import "common.proto";



// ============== 请求/响应 ==============

message UserOnlineRequest {
    string user_id = 1;
    string gate_id = 2;            // 用户连接的网关 ID
    string device_type = 3;        // windows, android, ios, web
    string device_id = 4;
}

message UserOfflineRequest {
    string user_id = 1;
    string gate_id = 2;
}

message RouteMessageRequest {
    string from_user_id = 1;
    string to_user_id = 2;
    bytes payload = 3;             // 消息内容（Protobuf 序列化）
    string msg_type = 4;           // 消息类型
}

message RouteMessageResponse {
    int32 code = 1;
    string message = 2;
    bool delivered = 3;            // 是否在线投递成功
}

message BroadcastRequest {
    repeated string user_ids = 1;  // 目标用户列表
    bytes payload = 2;
    string msg_type = 3;
}

message BroadcastResponse {
    int32 code = 1;
    string message = 2;
    int32 online_count = 3;        // 在线用户数
    int32 delivered_count = 4;     // 投递成功数
}

message GetUserStatusRequest {
    repeated string user_ids = 1;
}

message UserStatus {
    string user_id = 1;
    bool online = 2;
    string gate_id = 3;
    string device_type = 4;
    int64 last_active_at = 5;
}

message GetUserStatusResponse {
    int32 code = 1;
    string message = 2;
    repeated UserStatus statuses = 3;
}

message PushToUserRequest {
    string user_id = 1;
    string cmd = 2;                // 命令类型
    bytes payload = 3;             // 消息体
}

message KickUserRequest {
    string user_id = 1;
    string reason = 2;             // 踢出原因
}

// Gate 注册/心跳
message GateRegisterRequest {
    string gate_id = 1;
    string address = 2;            // Gate 的 gRPC 地址
    int32 current_connections = 3;
}

message GateHeartbeatRequest {
    string gate_id = 1;
    int32 current_connections = 2;
}

// ============== 客户端业务请求统一入口（Gate → Zone → 按 cmd 分发到各 System） ==============

message HandleClientRequestRequest {
    string conn_id = 1;            // Gate 侧连接 ID
    string user_id = 2;            // 已登录则有，用于鉴权与业务
    string cmd = 3;                // 如 auth.login、chat.send_message、friend.add
    bytes payload = 4;             // 业务请求体（由 cmd 决定解析方式）
    string request_id = 5;         // 客户端请求 ID，原样回填便于对账
    string token = 6;              // 已登录时 Gate 携带的 JWT，供 Zone 调业务服务时注入 metadata
}

message HandleClientRequestResponse {
    int32 code = 1;
    string message = 2;
    bytes payload = 3;             // 业务返回体（序列化后的 proto 或约定格式）
    string request_id = 4;         // 回填 request_id
}

// auth.login 请求体（cmd=auth.login 时 payload 为此消息序列化）
message AuthLoginPayload {
    string username = 1;
    string password = 2;
    string device_id = 3;
    string device_type = 4;
}

// auth.login 返回体（放入 HandleClientRequestResponse.payload）
message AuthLoginResponsePayload {
    bool success = 1;
    string user_id = 2;
    string token = 3;
    int64 expire_at = 4;
    string error = 5;
}

// auth.logout 请求体
message AuthLogoutPayload {
    string user_id = 1;
    string token = 2;
}

// auth.validate_token 请求体（仅 token）
message AuthValidateTokenPayload {
    string token = 1;
}

// auth.validate_token 返回体（payload 中仅 user_id，空表示无效）
message AuthValidateTokenResponsePayload {
    string user_id = 1;
}

// ---------- chat.* ----------
message ChatSendMessagePayload {
    string from_user_id = 1;
    string to_id = 2;
    int32 chat_type = 3;           // 1=私聊, 2=群聊
    string content = 4;
    string media_url = 5;
    string media_type = 6;
    string client_msg_id = 7;
    int64 file_size = 8;
    repeated string mentions = 9;  // @提醒的用户 ID 列表
    string reply_to_msg_id = 10;   // 回复的消息 ID
}

message ChatSendMessageResponsePayload {
    bool success = 1;
    string msg_id = 2;
    int64 timestamp = 3;
    string error = 4;
}

// 推送给在线用户的新消息体（cmd=chat.message 时 Gate 下发给客户端）
message ChatMessagePushPayload {
    string msg_id = 1;
    string from_user_id = 2;
    string to_id = 3;
    int32 chat_type = 4;           // 1=私聊, 2=群聊
    string content = 5;
    string media_url = 6;
    string media_type = 7;
    int64 timestamp = 8;
    repeated string mentions = 9;   // @提醒的用户 ID 列表
    string reply_to_msg_id = 10;   // 回复的消息 ID
}

message ChatRecallMessagePayload {
    string msg_id = 1;
    string user_id = 2;
}

// 标记已读请求（user_id 由 Zone 从 token 解析）
message ChatMarkReadPayload {
    string chat_id = 1;            // 私聊=对方 user_id，群聊=group_id
    int32 chat_type = 2;           // 1=私聊, 2=群聊
    string last_msg_id = 3;        // 已读到的最后一条消息 ID
}

// 拉取离线消息请求（user_id 由 Zone 从 token 解析）
message ChatPullOfflinePayload {
    int32 limit = 1;               // 单次拉取上限，默认 100，最大 200
    string cursor = 2;             // 分页游标
}

// 拉取离线消息返回体（payload）
message ChatPullOfflineResponsePayload {
    repeated ChatMessagePushPayload messages = 1;  // 离线消息列表（同在线推送格式）
    string next_cursor = 2;
    bool has_more = 3;
}

// ---------- friend.* ----------
message FriendAddPayload {
    string user_id = 1;
    string friend_id = 2;
    string remark = 3;
}

message FriendHandleRequestPayload {
    string user_id = 1;
    string request_id = 2;
    bool accept = 3;
}

message FriendRemovePayload {
    string user_id = 1;
    string friend_id = 2;
}

message FriendBlockPayload {
    string user_id = 1;
    string target_id = 2;
}

// ---------- group.* ----------
message GroupCreatePayload {
    string creator_id = 1;
    string group_name = 2;
    repeated string member_ids = 3;
}

message GroupCreateResponsePayload {
    bool success = 1;
    string group_id = 2;
    string error = 3;
}

message GroupDismissPayload {
    string group_id = 1;
    string operator_id = 2;
}

message GroupInviteMembersPayload {
    string group_id = 1;
    string inviter_id = 2;
    repeated string member_ids = 3;
}

message GroupRemoveMemberPayload {
    string group_id = 1;
    string operator_id = 2;
    string member_id = 3;
}

message GroupLeavePayload {
    string group_id = 1;
    string user_id = 2;
}

// ---------- file.* ----------
message FileGetUploadTokenPayload {
    string user_id = 1;
    string file_name = 2;
    int64 file_size = 3;
}

message FileGetUploadTokenResponsePayload {
    bool success = 1;
    string upload_token = 2;
    string upload_url = 3;
    int64 expire_at = 4;
    string error = 5;
}

message FileGetFileUrlPayload {
    string file_id = 1;
    string user_id = 2;
}

message FileGetFileUrlResponsePayload {
    bool success = 1;
    string file_url = 2;
    string file_name = 3;
    int64 file_size = 4;
    string content_type = 5;
    int64 expire_at = 6;
    string error = 7;
}

message FileDeletePayload {
    string file_id = 1;
    string user_id = 2;
}

// ============== 服务定义 ==============

service ZoneService {
    // 用户上线
    rpc UserOnline(UserOnlineRequest) returns (swift.common.CommonResponse);
    
    // 用户下线
    rpc UserOffline(UserOfflineRequest) returns (swift.common.CommonResponse);
    
    // 路由消息给单个用户
    rpc RouteMessage(RouteMessageRequest) returns (RouteMessageResponse);
    
    // 广播消息给多个用户
    rpc Broadcast(BroadcastRequest) returns (BroadcastResponse);
    
    // 获取用户在线状态
    rpc GetUserStatus(GetUserStatusRequest) returns (GetUserStatusResponse);
    
    // 主动推送消息给用户
    rpc PushToUser(PushToUserRequest) returns (swift.common.CommonResponse);
    
    // 踢用户下线
    rpc KickUser(KickUserRequest) returns (swift.common.CommonResponse);
    
    // Gate 注册
    rpc GateRegister(GateRegisterRequest) returns (swift.common.CommonResponse);
    
    // Gate 心跳
    rpc GateHeartbeat(GateHeartbeatRequest) returns (swift.common.CommonResponse);

    // 客户端业务请求统一入口：Gate 转发 cmd+payload，Zone 按 cmd 分发到对应 System，各 System 通过 gRPC 调后端
    rpc HandleClientRequest(HandleClientRequestRequest) returns (HandleClientRequestResponse);
}
